sycl/include/sycl/ext/oneapi/matrix/matrix-tensorcore.hpp

```diff
@@ -253,31 +396,73 @@ struct joint_matrix_mad_impl<
          LayoutC == sycl::ext::oneapi::experimental::matrix::matrix_layout::
                         col_major)>> {
   sycl::ext::oneapi::experimental::matrix::joint_matrix<
-      double, sycl::ext::oneapi::experimental::matrix::matrix_use::accumulator,
-      8, 8, LayoutC>
+      T2, sycl::ext::oneapi::experimental::matrix::matrix_use::accumulator, M,
+      N, LayoutC>
   mad(sycl::ext::oneapi::experimental::matrix::joint_matrix<
-          double, sycl::ext::oneapi::experimental::matrix::matrix_use::a, 8, 4,
+          T1, sycl::ext::oneapi::experimental::matrix::matrix_use::a, M, K,
           LayoutA>
           A,
       sycl::ext::oneapi::experimental::matrix::joint_matrix<
-          double, sycl::ext::oneapi::experimental::matrix::matrix_use::b, 4, 8,
+          T1, sycl::ext::oneapi::experimental::matrix::matrix_use::b, K, N,
           LayoutB>
           B,
       sycl::ext::oneapi::experimental::matrix::joint_matrix<
-          double,
-          sycl::ext::oneapi::experimental::matrix::matrix_use::accumulator, 8,
-          8, LayoutC>
+          T2, sycl::ext::oneapi::experimental::matrix::matrix_use::accumulator,
+          M, N, LayoutC>
           C) {
     sycl::ext::oneapi::experimental::matrix::joint_matrix<
-        double,
-        sycl::ext::oneapi::experimental::matrix::matrix_use::accumulator, 8, 8,
-        LayoutC>
+        T2, sycl::ext::oneapi::experimental::matrix::matrix_use::accumulator, M,
+        N, LayoutC>
         D;
 
 #ifdef __NVPTX__
 #ifdef __SYCL_DEVICE_ONLY__
-    __dmma_m8n8k4_mma_f64(D.data, A.data, B.data, C.data,
-                          get_layout_pair_id<LayoutA, LayoutB>(), 0);
+    if constexpr (M == 16 && N == 16 && K == 16) {
+      if constexpr (std::is_same<T1, int8_t>::value) {
+        __imma_m16n16k16_mma_s8(D.data, A.data, B.data, C.data,
+                                get_layout_pair_id<LayoutA, LayoutB>(), 0);
+      } else if constexpr (std::is_same<T1, uint8_t>::value) {
+        __imma_m16n16k16_mma_u8(D.data, A.data, B.data, C.data,
+                                get_layout_pair_id<LayoutA, LayoutB>(), 0);
+      } else if constexpr (std::is_same<T1, half>::value) {
+        __hmma_m16n16k16_mma_f32f32(D.data, A.data, B.data, C.data,
+                                    get_layout_pair_id<LayoutA, LayoutB>(), 0);
```

Comment:
A and B are always f16:  There is no ptx instruction for such a case.  

The __hmma_m16n16k16_mma_f16f16 instruction does exist.  This corresponds to the case where the "matrix_use::accumulator" matrices C (input) and D (output) are also fp16.  This case has not yet been added because it would require a change to the proposal where joint_matrix_mma takes an additional parameter that specifies the precision of the output matrix that would only be necessary for this mixed precision float case.
There are also the cases __hmma_m16n16k16_mma_f16f32 (C type f16 D type f32) and __hmma_m16n16k16_mma_f32f16.

A more pressing issue regarding the proposal for us is the one that would allow us to make use of the tf32 type:

The tf32 case was not added to this pull request because it requires a change to the joint_matrix_store/joint_matrix_load interface so that loading/storing the accumulator matrix for the tf32 case can be distinguished from the other floating point cases. The tf32 case uses a different instruction to perform these loads/stores. This instruction currently cannot be distinguished from the other fp32 accumulator load/store cases since they all use the same matrix dimensions and fp32 type for the matrix elements. The most straightforward solution from our point of view is to add an additional template parameter to joint_matrix_store/joint_matrix_load corresponding to the "K" dimension that is used to specify the shapes of A/B matrices that may be used in a MMA with the Accumulator matrix that is being loaded/stored. In this case joint_matrix_load/joint_matrix_store would take three template parameters instead of two, corresponding to the full set of values that specify the shapes of the A/B/Accumulator matrices: M, N, K. How would this change be from your point of view?

Reply:
I think __hmma_m16n16k16_mma_f16f16 can be added with the current interface by checking if C is fp16. If C is fp32, __hmma_m16n16k16_mma_f32f32 is generated.
BTW, you should add a check that T2 is fp32 to line 427.

Reply:
tf32 is a separate issue. I think the current direction is to introduce a special type like we did for bf16 or something along that. So I am not sure whether introducing extra complexity to the matrix interface is the right thing to do at this point.

One immediate workaround can be to treat fp32 as tf32 for storage, the same way we are treating unsigned short as bf16. 
We add a comment about this stating that this is a workaround while tf32 type is being baked.

What do you think?

Reply:
So the accumulator matrix for tf32 is always fp32 already - this isn't the issue.  If you look here: https://docs.nvidia.com/cuda/parallel-thread-execution/index.html#warp-level-matrix-instructions-wmma-ld
You can see that there is a different instruction for loading the accumulator matrix for the tf32 type compared to loading the accumulator matrix for the bf16 MMA or the mixed precision float MMA:  The tf32 MMA uses the following instruction:

`wmma.load.c.sync.aligned.layout..m16n16k8{.ss}.f32`

whereas the other two cases use:

`wmma.load.c.sync.aligned.layout..m16n16k8{.ss}.f32`

There is no way to distinguish between these two cases using the current interface of joint_matrix_load, since the values for NumRows=16, NumCols=16 and T=f32 are identical in both cases.  An obvious resolution is to add a parameter for k=8/16, which is what the Cuda runtime does.  Another reason for introducing this extra parameter is that it adds semantics that joint_matrix_load/joint_matrix_store/joint_matrix_mma should always be used together with matching values for M,N,K, but an argument against is that it is more complicated since there is an additional parameter that isn't used most of the time.  I wanted to check with you whether the Intel backend might also have a similar case to be implemented in the future that might require the additional parameter corresponding to K, and more generally what you think about adding the K parameter.  Otherwise I think we would have to add an additional runtime function for the tf32 type for the cuda backend: e.g. "joint_matrix_load_tf32_cuda" or similar.

Reply:
> I think __hmma_m16n16k16_mma_f16f16 can be added with the current interface by checking if C is fp16. If C is fp32, __hmma_m16n16k16_mma_f32f32 is generated. BTW, you should add a check that T2 is fp32 to line 427.

Yeah I should be able to add __hmma_m16n16k16_mma_f16f16 with the current interface.  I wanted to also check first whether you also have any cases where the output matrix "D" in joint_matrix_mma differs from the "C" matrix?
SImilar to e.g.:
`__hmma_m16n16k16_mma_f16f32` or
`__hmma_m16n16k16_mma_f32f16`

If you do then it might make sense to update the interface accordingly at this point.  If not, since it is not really a priority for us to support these cases I can just add the __hmma_m16n16k16_mma_f16f16 case using the current joint_matrix_mma interface which assumes the types of "C" and "D" are both "T2". As I mentioned earlier it is always the case for cuda that Matrices "A" and "B" are "T1", so that part is fine for us.